File: examples/FSharp/SequentialSteps.fs
module SequentialSteps

open NBomber
open NBomber.Contracts
open NBomber.Http
open NBomber.Http.FSharp
open NBomber.FSharp
open NBomber.Plugins.Network.Ping

[<CLIMutable>]
type UserData = {
    UserId: int
    Id: int
    Title: string
    Completed: bool
}

let run () =

    use httpClient = Http.createDefaultClient()

    Scenario.create("http_scenario", fun context -> task {

        let! step1 = Step.run("step_1", context, fun _ -> task {

            let! response =
                Http.createRequest "GET" "https://nbomber.com"
                |> Http.send httpClient

            return response
        })

        // example of WithJsonBody
        let! step2 = Step.run("step_2", context, fun _ -> task {

            let user = {| Id = "1"; Name = "Test Name" |}

            let! response =
                Http.createRequest "GET" "https://nbomber.com"
                |> Http.withJsonBody user
                |> Http.send httpClient

            return response
        })

        // example of Http.sendTyped
        let! step3 = Step.run("step_3", context, fun _ -> task {

            let! response =
                Http.createRequest "GET" "https://jsonplaceholder.typicode.com/todos/1"
                |> Http.sendTyped<UserData> httpClient

            return response
        })

        return Response.ok()
    })
    |> Scenario.withWarmUpDuration(seconds 3)
    |> Scenario.withLoadSimulations [Inject(rate = 5, interval = seconds 1, during = minutes 1)]
    |> NBomberRunner.registerScenario
    |> NBomberRunner.withWorkerPlugins [
        new PingPlugin(PingPluginConfig.createDefault "nbomber.com")
        new HttpMetricsPlugin()
    ]
    |> NBomberRunner.run
    |> ignore

File: examples/CSharp/SequentialSteps.cs
using System;
using System.Net.Http;
using System.Text.Json;
using NBomber.CSharp;
using NBomber.Http;
using NBomber.Http.CSharp;

namespace CSharp;

class SequentialSteps
{
    public void Run()
    {
        // sets global JsonSerializerOptions to use CamelCase naming
        Http.GlobalJsonSerializerOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        };

        using var httpClient = new HttpClient();

        var scenario = Scenario.Create("http_scenario", async context =>
        {
            var step1 = await Step.Run("step_1", context, async () =>
            {
                var request =
                    Http.CreateRequest("GET", "https://nbomber.com")
                        .WithHeader("Accept", "application/json")
                        .WithBody(new StringContent("{ some JSON }"));

                var response = await Http.Send(httpClient, request);

                return response;
            });

            // example of WithJsonBody<T>(data)
            // populates request body by serializing data record to JSON format. Also, it adds HTTP header: "Accept": "application/json".

            var step2 = await Step.Run("step_2", context, async () =>
            {
                var user = new UserData { UserId = 1, Title = "anton" };

                var request = Http.CreateRequest("GET", "https://nbomber.com")
                                  .WithJsonBody(user);

                var response = await Http.Send(httpClient, request);

                return response;
            });

            // example of Http.Send<TResponse>
            // send request and deserialize HTTP response body to JSON format

            var step3 = await Step.Run("step_3", context, async () =>
            {
                var request =
                    Http.CreateRequest("GET", "https://jsonplaceholder.typicode.com/todos/1")
                        .WithHeader("Accept", "application/json");

                var response = await Http.Send<UserData>(httpClient, request);

                // user: UserData type
                var user = response.Payload.Value;
                var userId = response.Payload.Value.UserId;

                return response;
            });

            return Response.Ok();
        })
        .WithoutWarmUp()
        .WithLoadSimulations(Simulation.Inject(rate: 5, interval: TimeSpan.FromSeconds(1), during: TimeSpan.FromSeconds(30)));

        NBomberRunner
            .RegisterScenarios(scenario)
            .WithWorkerPlugins(new HttpMetricsPlugin())
            .Run();
    }
}
